crates/astria-celestia-client/src/client.rs

```diff
@@ -0,0 +1,306 @@
+        SequencerNamespaceData,
+        SignedNamespaceData,
+    },
+    RollupNamespaceData,
+};
+
+impl CelestiaClientExt for jsonrpsee::http_client::HttpClient {}
+
+#[derive(Debug, thiserror::Error)]
+pub enum SubmitSequencerBlocksError {
+    #[error("failed assembling blob for block at index `{index}`")]
+    AssembleBlobs {
+        source: BlobAssemblyError,
+        index: usize,
+    },
+    #[error("the JSONRPC call failed")]
+    JsonRpc(#[source] jsonrpsee::core::Error),
+}
+
+pub struct BadBlob {
+    pub reason: BadBlobReason,
+    pub commitment: Commitment,
+}
+
+pub enum BadBlobReason {
+    Deserialization(serde_json::Error),
+    WrongNamespace(Namespace),
+}
+
+pub struct GetSequencerDataResponse {
+    pub height: u64,
+    pub namespace: Namespace,
+    pub datas: Vec<SignedNamespaceData<SequencerNamespaceData>>,
+    pub bad_blobs: Vec<BadBlob>,
+}
+
+pub struct GetRollupDataResponse {
+    pub height: u64,
+    pub namespace: Namespace,
+    pub datas: Vec<SignedNamespaceData<RollupNamespaceData>>,
+    pub bad_blobs: Vec<BadBlob>,
+}
+
+#[async_trait]
+pub trait CelestiaClientExt: BlobClient {
+    /// Fetch sequencer data at the given height with the provided namespace.
+    ///
+    /// Returns successfully deserialized blobs in the `.datas` field. The
+    /// `.bad_blobs` field contains the celestia commitment for each blob
+    /// that could not be turned into sequencer data and the reason for it.
+    ///
+    /// # Errors
+    ///
+    /// Fails if the underlying `blob.GetAll` JSONRPC failed.
+    async fn get_sequencer_data(
+        &self,
+        height: u64,
+        namespace: Namespace,
+    ) -> Result<GetSequencerDataResponse, jsonrpsee::core::Error> {
+        let blobs = self.blob_get_all(height, &[namespace]).await?;
+
+        let mut datas = Vec::new();
+        let mut bad_blobs = Vec::new();
+        for blob in blobs {
+            if blob.namespace != namespace {
+                bad_blobs.push(BadBlob {
+                    reason: BadBlobReason::WrongNamespace(blob.namespace),
+                    commitment: blob.commitment,
+                });
+            }
+            match serde_json::from_slice(&blob.data) {
+                Ok(data) => datas.push(data),
+                Err(err) => bad_blobs.push(BadBlob {
+                    reason: BadBlobReason::Deserialization(err),
+                    commitment: blob.commitment,
+                }),
+            }
+        }
+
+        Ok(GetSequencerDataResponse {
+            height,
+            namespace,
+            datas,
+            bad_blobs,
+        })
+    }
+
+    /// Returns the rollup data for a given rollup namespace at a given height, if it exists.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if:
+    /// + the verification key could not be constructed from the data stored in `namespace_data`;
+    /// + the RPC to fetch the blobs failed.
+    async fn get_rollup_data_matching_sequencer_data(
```

Comment:
Flagging this function for review. This used to return just a single roll-up namespace making certain assumptions about how blobs are submitted to celestia.

Now this is more generic, allowing multiple rollup blobs for one roll-up.

Reply:
yeah this seems fine, it should be able to return multiple blobs (since sequencer will be submitting multiple blobs per celestia block anyways)